File: packages/shared/src/payments.test.ts
import { describe, expect, it } from "vitest";

import { EN as WEB_EN, FR_CA as WEB_FR } from "../../../apps/web/src/i18n/catalog";

import {
  PAYMENT_MAX_CENTS,
  PAYMENT_MIN_CENTS,
  paymentAmountProblem,
  paymentRequestCancellable,
  paymentRequestLabel,
  paymentRequestState,
  payoutReadiness,
  payoutReadinessCopy,
  payoutRequirementCopy,
} from "./payments";
import { paymentAmountProblemCopy, paymentRequestSms } from "./payments-copy";

/** #228 — resolve a catalogue key the way the three strips do. */
function look(table: unknown, key: string): string {
  const [section, name] = key.split(".");
  const value = (table as Record<string, Record<string, string>>)[section]?.[
    name
  ];
  if (typeof value !== "string") throw new Error(`no entry for ${key}`);
  return value;
}

const sayEn = (key: string): string => look(WEB_EN, key);
const sayFr = (key: string): string => look(WEB_FR, key);

/**
 * #224 — the vectors the Kotlin and Swift ports are checked against.
 *
 * Three clients hand-port this file, and #548 is the standing lesson: a rule
 * written three times drifts. Every case below is one a port has to reproduce,
 * and the state-precedence block is the one that decides whether a crew chases
 * a customer for money they already paid.
 */
describe("paymentRequestState", () => {
  it("reads a plain open request as waiting", () => {
    expect(paymentRequestState({ status: "requested" })).toBe("requested");
  });

  it("reads paid from either the status or the stamp", () => {
    expect(paymentRequestState({ status: "paid" })).toBe("paid");
    expect(
      paymentRequestState({ status: "requested", paid_at: "2026-08-01T00:00:00Z" }),
    ).toBe("paid");
  });

  it("puts a dispute above a refund, because a chargeback needs somebody", () => {
    expect(
      paymentRequestState({
        status: "paid",
        paid_at: "2026-08-01T00:00:00Z",
        refunded_at: "2026-08-02T00:00:00Z",
        disputed_at: "2026-08-03T00:00:00Z",
      }),
    ).toBe("disputed");
  });

  it("puts a refund above paid", () => {
    expect(
      paymentRequestState({
        status: "paid",
        paid_at: "2026-08-01T00:00:00Z",
        refunded_at: "2026-08-02T00:00:00Z",
      }),
    ).toBe("refunded");
  });

  /**
   * THE case. A cancelled request whose link somehow still took a payment must
   * read as PAID — the money is real, and telling a crew otherwise is how a
   * customer gets chased for a bill they already settled.
   */
  it("reads a cancelled-but-paid request as paid", () => {
    expect(
      paymentRequestState({
        status: "cancelled",
        paid_at: "2026-08-01T00:00:00Z",
      }),
    ).toBe("paid");
  });

  it("reads cancelled and expired straight through", () => {
    expect(paymentRequestState({ status: "cancelled" })).toBe("cancelled");
    expect(paymentRequestState({ status: "expired" })).toBe("expired");
  });
});

describe("paymentRequestLabel", () => {
  it("gives every state exactly one word", () => {
    const states = [
      "requested",
      "paid",
      "refunded",
      "disputed",
      "cancelled",
      "expired",
    ] as const;
    const keys = states.map((state) => paymentRequestLabel(state));
    expect(keys).toEqual([
      "payments.stateWaiting",
      "payments.statePaid",
      "payments.stateRefunded",
      "payments.stateDisputed",
      "payments.stateCancelled",
      "payments.stateExpired",
    ]);
    // No two states share a word: a reader must never have to guess which of
    // two situations a row is in. Checked on the WORDS, in both languages —
    // six distinct keys resolving to five distinct words is the same defect
    // wearing a disguise the key comparison cannot see.
    for (const table of [WEB_EN, WEB_FR]) {
      const words = keys.map((key) => look(table, key));
      expect(new Set(words).size).toBe(words.length);
      for (const word of words) expect(word.length).toBeGreaterThan(0);
    }
  });
});

describe("paymentRequestCancellable", () => {
  it("offers cancel only while the request is genuinely open", () => {
    expect(paymentRequestCancellable({ status: "requested" })).toBe(true);
    expect(paymentRequestCancellable({ status: "paid" })).toBe(false);
    expect(paymentRequestCancellable({ status: "cancelled" })).toBe(false);
    // Expired is already dead. A Cancel on it would be a tap that does nothing,
    // which reads as a broken button rather than a settled state.
    expect(paymentRequestCancellable({ status: "expired" })).toBe(false);
  });
});

describe("paymentAmountProblem", () => {
  it("accepts an ordinary trade amount", () => {
    expect(paymentAmountProblem(25_000)).toBeNull();
  });

  it("refuses below Stripe's own floor", () => {
    expect(paymentAmountProblem(PAYMENT_MIN_CENTS - 1)).toBe("too_small");
    expect(paymentAmountProblem(PAYMENT_MIN_CENTS)).toBeNull();
  });

  it("refuses the missed decimal", () => {
    expect(paymentAmountProblem(PAYMENT_MAX_CENTS + 1)).toBe("too_large");
    expect(paymentAmountProblem(PAYMENT_MAX_CENTS)).toBeNull();
  });

  it("refuses a fractional cent", () => {
    expect(paymentAmountProblem(1000.5)).toBe("not_whole");
  });

  it("states the bound in the reader's own currency", () => {
    expect(paymentAmountProblemCopy("too_small", "cad", sayEn)).toContain("$1");
    expect(paymentAmountProblemCopy("too_large", "usd", sayEn)).toContain("$25,000");
  });

  it("#228: states it in the reader's own LANGUAGE too", () => {
    // The figure is formatted by `formatMoney` and the sentence around it is
    // a key. Both halves have to survive: a French refusal that lost the
    // amount would tell somebody they are outside a limit without saying
    // which, and that is the one fact they need to fix it.
    const small = paymentAmountProblemCopy("too_small", "cad", sayFr);
    expect(small).toMatch(/^Le plus petit paiement/);
    expect(small).toContain("$1");
    expect(small, "a variable survived the fill").not.toMatch(/\{/);

    const large = paymentAmountProblemCopy("too_large", "usd", sayFr);
    expect(large).toContain("$25,000");
    expect(large).not.toBe(paymentAmountProblemCopy("too_large", "usd", sayEn));
  });
});

describe("paymentRequestSms", () => {
  const text = paymentRequestSms({
    businessName: "Northline Plumbing",
    amountCents: 25_000,
    currency: "usd",
    description: "Deposit for Tuesday",
    url: "https://app.loonext.com/pay/abc",
  });

  it("leads with the business, then the amount", () => {
    expect(text.startsWith("Northline Plumbing: $250 for Deposit for Tuesday.")).toBe(
      true,
    );
  });

  it("puts the link last and on its own line, so every phone linkifies it", () => {
    const lines = text.split("\n");
    expect(lines[lines.length - 1]).toBe("https://app.loonext.com/pay/abc");
  });

  it("says nothing a spam filter or a homeowner reads as pressure", () => {
    for (const banned of ["click here", "urgent", "immediately", "act now"]) {
      expect(text.toLowerCase()).not.toContain(banned);
    }
  });
});

describe("payoutReadiness", () => {
  const base = {
    connected: true,
    charges_enabled: false,
    details_submitted: false,
    disabled_reason: null,
    requirements_due: [],
  };

  it("is not_connected with no account at all", () => {
    expect(payoutReadiness(null)).toBe("not_connected");
    expect(payoutReadiness({ ...base, connected: false })).toBe("not_connected");
  });

  /**
   * `charges_enabled` decides, and it decides ALONE. A restricted account with
   * charges somehow still on is ready — because Stripe says it can take the
   * card, and Stripe is the only thing that knows.
   */
  it("is ready whenever Stripe says charges are enabled", () => {
    expect(payoutReadiness({ ...base, charges_enabled: true })).toBe("ready");
    expect(
      payoutReadiness({
        ...base,
        charges_enabled: true,
        disabled_reason: "requirements.past_due",
      }),
    ).toBe("ready");
  });

  it("names the reason when charges are off", () => {
    expect(
      payoutReadiness({ ...base, disabled_reason: "requirements.past_due" }),
    ).toBe("restricted");
    expect(payoutReadiness(base)).toBe("onboarding_incomplete");
    // Everything handed over, nothing outstanding, charges still off: Stripe is
    // looking at it, and there is nothing for the owner to do.
    expect(payoutReadiness({ ...base, details_submitted: true })).toBe(
      "pending_verification",
    );
  });
});

describe("payoutReadinessCopy", () => {
  it("offers an action for every state a person can act on", () => {
    for (const readiness of [
      "not_connected",
      "onboarding_incomplete",
      "restricted",
      "ready",
    ] as const) {
      expect(payoutReadinessCopy(readiness).action).toBeTruthy();
    }
  });

  it("offers NO action while Stripe is verifying, and says so", () => {
    const copy = payoutReadinessCopy("pending_verification");
    expect(copy.action).toBeNull();
    expect(copy.detail).toContain("nothing for you to do");
  });

  it("never promises the platform takes a cut", () => {
    expect(payoutReadinessCopy("not_connected").detail).toContain(
      "take nothing on top",
    );
  });
});

describe("payoutRequirementCopy", () => {
  /*
   * #228 — resolved through the catalogue the card reads, not against a copy
   * of the sentence held here. An assertion holding its own copy keeps passing
   * after the catalogue moves underneath it, and the screen becomes the only
   * place the two disagree.
   */
  const say = (key: keyof typeof WEB_EN.payments): string => WEB_EN.payments[key];

  it("turns Stripe's identifiers into words a plumber reads", () => {
    expect(payoutRequirementCopy("external_account")).toEqual({
      key: "payments.reqBankAccount",
      literal: null,
    });
    expect(say("reqBankAccount")).toBe("Your bank account details");

    expect(payoutRequirementCopy("individual.verification.document")).toEqual({
      key: "payments.reqOwnerId",
      literal: null,
    });
    expect(say("reqOwnerId")).toBe("Photo ID for the business owner");
  });

  it("says all twelve in French, which is the half nobody re-reads", () => {
    // A key present in English and missing in French falls back to English by
    // design, so this never surfaces as a broken screen — it surfaces as a
    // French owner reading an English list of what Stripe still wants.
    for (const requirement of [
      "external_account",
      "business_profile.url",
      "business_profile.mcc",
      "individual.verification.document",
      "individual.verification.additional_document",
      "individual.id_number",
      "individual.address.line1",
      "individual.dob.day",
      "company.tax_id",
      "company.verification.document",
      "tos_acceptance.date",
      "representative.verification.document",
    ]) {
      const { key } = payoutRequirementCopy(requirement);
      expect(key, requirement).not.toBeNull();
      const name = key!.slice("payments.".length) as keyof typeof WEB_FR.payments;
      expect(typeof WEB_FR.payments[name], requirement).toBe("string");
      expect(WEB_FR.payments[name], requirement).not.toBe(WEB_EN.payments[name]);
    }
  });

  /**
   * The fallback matters more than the table. Stripe adds requirement
   * identifiers without telling anybody, and an outstanding requirement nobody
   * can see is the state where an owner concludes the product is broken — so an
   * unknown identifier is made readable rather than dropped.
   */
  it("renders an unknown identifier rather than swallowing it", () => {
    // And as a LITERAL rather than a key, because inventing French for a
    // requirement we do not recognise would be inventing the requirement.
    expect(payoutRequirementCopy("company.owners_provided")).toEqual({
      key: null,
      literal: "Owners provided",
    });
    expect(payoutRequirementCopy("individual.political_exposure")).toEqual({
      key: null,
      literal: "Political exposure",
    });
  });

  it("sets exactly one of the two, so a caller cannot read the wrong field", () => {
    for (const requirement of ["external_account", "company.owners_provided"]) {
      const copy = payoutRequirementCopy(requirement);
      expect((copy.key === null) !== (copy.literal === null), requirement).toBe(true);
    }
  });
});

File: apps/android/app/src/test/kotlin/com/loonext/android/features/payments/PaymentsTest.kt
package com.loonext.android.features.payments

import com.loonext.android.features.settings.BillingCurrency
import com.loonext.android.features.settings.formatMoney
import java.time.Instant
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #224 — the halves of the port the generated vectors deliberately do not
 * cover.
 *
 * `scripts/generate-parity-vectors.mjs` pins the RULES (state, bounds,
 * readiness) and excludes copy and presentation on purpose. What is left is
 * still where a hand-port goes wrong, and each test below names the failure it
 * exists to prevent rather than the function it calls.
 */
class PaymentsTest {

    private val usd = BillingCurrency.USD
    private val cad = BillingCurrency.CAD

    // --- parseAmountToCents -------------------------------------------------

    @Test
    fun `a typed amount becomes cents through integers`() {
        assertEquals(25_000, Payments.parseAmountToCents("250"))
        assertEquals(25_050, Payments.parseAmountToCents("250.50"))
        // One decimal place means TENTHS of a dollar, not cents. "250.5" is
        // $250.50, and a port that read the fraction as-is would bill $250.05.
        assertEquals(25_050, Payments.parseAmountToCents("250.5"))
        assertEquals(100, Payments.parseAmountToCents("1"))
        assertEquals(1, Payments.parseAmountToCents("0.01"))
    }

    @Test
    fun `19 dollars 99 is exactly 1999 cents`() {
        // THE case the arithmetic exists for. 19.99 as a double times 100 is
        // 1998.9999999999998, and rounding that is a coin-flip nobody should
        // take with somebody's bill.
        assertEquals(1999, Payments.parseAmountToCents("19.99"))
        // Escaped, because a bare `$` opens a string template in Kotlin — the
        // same class of hand-port trap as `\b` being a backspace here.
        assertEquals(1999, Payments.parseAmountToCents("\$19.99"))
    }

    @Test
    fun `decoration a thumb adds is forgiven`() {
        // A dollar sign and a thousands separator are what somebody types
        // without thinking. Refusing them would be pedantry at the one keyboard
        // this feature is used from.
        assertEquals(120_000, Payments.parseAmountToCents("\$1,200"))
        assertEquals(25_000, Payments.parseAmountToCents(" 250 "))
    }

    @Test
    fun `anything that is not a number is refused rather than guessed`() {
        // A silently-misread amount is the one error this feature cannot afford,
        // so the shape is strict even though the decoration is not.
        for (input in listOf("", "   ", "abc", "25.999", "1.2.3", "-50", "1e3", ".5", "250-")) {
            assertNull("'$input' should not parse", Payments.parseAmountToCents(input))
        }
    }

    @Test
    fun `a fat-fingered number never wraps into a small one`() {
        // The overflow guard is load-bearing, not defensive: a value that
        // wrapped and happened to land between the floor and the ceiling would
        // be a bill for a figure nobody typed. Refusing is the only safe answer,
        // and the ceiling check downstream never sees it.
        assertNull(Payments.parseAmountToCents("999999999999999999999999999999"))
        assertNull(Payments.parseAmountToCents("99999999"))
        // Still a number, still refused — but by the CEILING, with a sentence.
        assertEquals(Payments.MAX_CENTS + 100, Payments.parseAmountToCents("25001"))
        assertEquals(
            PaymentAmountProblem.TOO_LARGE,
            Payments.amountProblem(Payments.parseAmountToCents("25001")!!),
        )
    }

    @Test
    fun `a fraction of a cent cannot be typed at all`() {
        // Why the Int-typed [Payments.amountProblem] can never answer NOT_WHOLE
        // on this client: the parse refuses a third decimal place before
        // anything downstream sees it.
        assertNull(Payments.parseAmountToCents("1.234"))
        assertNull(Payments.parseAmountToCents("10.005"))
    }

    // --- the copy that carries money ---------------------------------------

    @Test
    fun `a refusal names the bound in the reader's own money`() {
        // Not a typed price. A Canadian workspace's Stripe account settles in
        // CAD, and "$1" typed as a literal would be the #522 defect on the one
        // sentence that tells somebody what they may charge.
        assertEquals(
            "The smallest payment we can take is ${formatMoney(Payments.MIN_CENTS, cad)}.",
            Payments.amountProblemCopy(PaymentAmountProblem.TOO_SMALL, cad),
        )
        assertEquals(
            "The largest payment we can take by text is " +
                "${formatMoney(Payments.MAX_CENTS, usd)}.",
            Payments.amountProblemCopy(PaymentAmountProblem.TOO_LARGE, usd),
        )
    }

    @Test
    fun `every amount problem has a sentence, and none of it is a code`() {
        for (problem in PaymentAmountProblem.entries) {
            val copy = Payments.amountProblemCopy(problem, usd)
            assertTrue("$problem has no copy", copy.length > 20)
            assertTrue("$problem reads like a code", copy.contains(" "))
            assertEquals(copy.trim(), copy)
        }
    }

    @Test
    fun `every state has exactly one word`() {
        for (state in PaymentState.entries) {
            val label = Payments.label(state)
            assertTrue("$state has no label", label.isNotBlank())
            assertFalse("$state's label is a sentence", label.contains(" "))
        }
    }

    // --- the SMS ------------------------------------------------------------

    @Test
    fun `the text names the business first and the link last`() {
        val text = Payments.requestSms(
            businessName = "  Maple Plumbing  ",
            amountCents = 25_000,
            currency = cad,
            description = "  Deposit  ",
            url = "https://app.loonext.com/pay/abc123",
        )
        assertEquals(
            "Maple Plumbing: ${formatMoney(25_000, cad)} for Deposit.\n" +
                "Pay securely here:\nhttps://app.loonext.com/pay/abc123",
            text,
        )
        // A payment link from an unnamed sender is a phishing text, and the
        // customer is right to think so.
        assertTrue(text.startsWith("Maple Plumbing:"))
        // On its own line, so every phone linkifies all of it.
        assertTrue(text.endsWith("\nhttps://app.loonext.com/pay/abc123"))
        // Not what a carrier's spam filter is looking for.
        assertFalse(text.lowercase().contains("click here"))
        assertFalse(text.contains("!"))
    }

    // --- Stripe requirements ------------------------------------------------

    @Test
    fun `a Stripe requirement identifier becomes something a plumber can read`() {
        assertEquals(
            "Photo ID for the business owner",
            Payments.requirementCopy("individual.verification.document"),
        )
        assertEquals("Your bank account details", Payments.requirementCopy("external_account"))
    }

    @Test
    fun `an identifier nobody wrote copy for is still shown`() {
        // Dropping it would leave an owner with an outstanding requirement they
        // cannot see, which is the state where somebody concludes the product is
        // broken. The fallback strips whose-document-is-it and the separators.
        assertEquals(
            "Political exposure",
            Payments.requirementCopy("individual.political_exposure"),
        )
        assertEquals("Ownership declaration", Payments.requirementCopy("ownership_declaration"))
        // Only a LEADING owner prefix goes; one in the middle is part of the name.
        assertEquals(
            "Directors individual name",
            Payments.requirementCopy("directors.individual.name"),
        )
    }

    @Test
    fun `an empty identifier does not crash the settings card`() {
        assertEquals("", Payments.requirementCopy(""))
    }

    // --- what the strip shows ----------------------------------------------

    @Test
    fun `a live request is always worth showing`() {
        // However old. A request nobody has paid is the whole point of the strip.
        val ancient = Instant.parse("2020-01-01T00:00:00Z").toString()
        assertTrue(
            Payments.isWorthShowing(PaymentState.REQUESTED, ancient, null, NOW),
        )
    }

    @Test
    fun `a settled request drops off after a week`() {
        val sixDays = Instant.ofEpochMilli(NOW - 6L * DAY).toString()
        val eightDays = Instant.ofEpochMilli(NOW - 8L * DAY).toString()
        assertTrue(Payments.isWorthShowing(PaymentState.PAID, sixDays, sixDays, NOW))
        assertFalse(Payments.isWorthShowing(PaymentState.PAID, eightDays, eightDays, NOW))
        // Dated from when it was PAID, not from when it was asked for: a deposit
        // requested in March and paid yesterday is yesterday's news.
        assertTrue(Payments.isWorthShowing(PaymentState.PAID, eightDays, sixDays, NOW))
    }

    @Test
    fun `a row we cannot date is shown rather than hidden`() {
        // The rows that can be settled include DISPUTED, and a chargeback the
        // crew never sees is the one failure this strip exists to prevent.
        assertTrue(Payments.isWorthShowing(PaymentState.DISPUTED, "not-a-date", null, NOW))
        assertTrue(Payments.isWorthShowing(PaymentState.PAID, null, null, NOW))
    }

    // --- the account model --------------------------------------------------

    @Test
    fun `a readiness this build has never heard of falls back to the booleans`() {
        // Both constants would be wrong in a way that costs somebody money:
        // NOT_CONNECTED hides the ask from a workspace that can charge, READY
        // offers it to one that cannot and turns every send into a 409.
        val account = PayoutAccount(
            connected = true,
            readiness = "awaiting_something_invented_next_quarter",
            charges_enabled = true,
            details_submitted = true,
        )
        assertEquals(PayoutReadiness.READY, account.state)
        assertTrue(account.canCharge)

        val notYet = account.copy(charges_enabled = false)
        assertEquals(PayoutReadiness.PENDING_VERIFICATION, notYet.state)
        assertFalse(notYet.canCharge)
    }

    @Test
    fun `the server's own readiness is used when we recognise it`() {
        val account = PayoutAccount(
            connected = true,
            readiness = PayoutReadiness.RESTRICTED.wire,
            charges_enabled = false,
            details_submitted = true,
            disabled_reason = "requirements.past_due",
        )
        assertEquals(PayoutReadiness.RESTRICTED, account.state)
    }

    @Test
    fun `the amount is quoted in what the account settles in`() {
        // Stripe's default_currency first, the account's COUNTRY second, never a
        // platform default. A Canadian account quoted in USD would settle at
        // Stripe's conversion rate and the business would receive less than the
        // number they typed.
        assertEquals(
            BillingCurrency.CAD,
            PayoutAccount(connected = true, currency = "cad", country = "US").billingCurrency,
        )
        assertEquals(
            BillingCurrency.CAD,
            PayoutAccount(connected = true, currency = null, country = "CA").billingCurrency,
        )
        assertEquals(
            BillingCurrency.USD,
            PayoutAccount(connected = true, currency = null, country = null).billingCurrency,
        )
    }

    @Test
    fun `a request derives its own state from its own timestamps`() {
        // Derived rather than decoded from the wire `state`, so a row painted
        // from the process cache before any fetch says the same word the API
        // would. THE case: cancelled, then paid anyway.
        val row = PaymentRequest(
            id = "1",
            status = PaymentStatus.CANCELLED,
            paid_at = "2026-08-01T00:00:00Z",
            amount_cents = 25_000,
            currency = "cad",
        )
        assertEquals(PaymentState.PAID, row.state)
        assertFalse(Payments.cancellable(row.state))
        assertEquals(BillingCurrency.CAD, row.money)
    }

    private companion object {
        const val DAY = 24L * 60 * 60 * 1000
        /** A fixed instant: a test that computes its window from the wall clock
         *  is a test that starts failing on a date nobody chose. */
        val NOW = Instant.parse("2026-08-10T12:00:00Z").toEpochMilli()
    }
}
